api-v2/src/plan/plan.service.ts

```diff
@@ -0,0 +1,77 @@
+import { Injectable } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Student } from 'src/student/entities/student.entity';
+import { StudentService } from 'src/student/student.service';
+import { Repository } from 'typeorm';
+import { CreatePlanDto } from './dto/create-plan.dto';
+import { UpdatePlanDto } from './dto/update-plan.dto';
+import { Plan } from './entities/plan.entity';
+import { PlanNotFoundError } from './plan-not-found.error';
+
+@Injectable()
+export class PlanService {
+  constructor(
+    @InjectRepository(Plan)
+    private planRepository: Repository<Plan>,
+  ) {}
+
+  create(createPlanDto: CreatePlanDto, student: Student): Promise<Plan> {
+    const newPlan = this.planRepository.create({ ...createPlanDto, student });
+
+    try {
+      return this.planRepository.save(newPlan);
+    } catch (_error) {
+      throw new Error('Something went wrong when creating the new Plan');
+    }
+  }
+
+  async findOne(id: number): Promise<Plan> {
+    const plan = this.planRepository.findOne({
+      where: { id },
+      relations: ['student'],
+    });
+
+    if (!plan) {
+      throw new PlanNotFoundError(id);
+    }
+
+    return plan;
+  }
+
+  async isPlanOwnedByStudent(
+    planId: number,
+    loggedInStudent: Student,
+  ): Promise<boolean> {
+    try {
+      const { student: planOwner } = await this.findOne(planId);
+      return StudentService.compareStudents(planOwner, loggedInStudent);
+    } catch (_error) {
+      /**
+       * findOne failed probably since the plan doesn't exist. A plan that doesn't
+       * exist isn't owner by the anyone so returning false.
+       */
+      return false;
+    }
+  }
+
+  async update(id: number, updatePlanDto: UpdatePlanDto) {
+    const updateResult = await this.planRepository.update(id, updatePlanDto);
+
+    if (updateResult.affected === 0) {
+      // no plan was updated, which implies a plan with the given id wasn't found
+      throw new PlanNotFoundError(id);
+    }
```

Comment:
seeing all these error cases is making me want to write test cases to cover them all lol

Reply:
lmaoo yea, writing endpoint tests shouldn't be too much work. I've manually tested every edge case and that definitely isn't the way to do things lol.

Maybe we can make this a part of this PR or split it out into its own ticket if someone wants to gain some experience with nest, jest, and supertest. The good thing is that we're merging into a feature branch and not main. But if you'd like to see them in this PR, can totally make it happen!

Reply:
nah we do this later 🤪